mlrun/api/schemas/model_endpoints.py

```diff
@@ -1,106 +1,227 @@
+            )
+            self.metadata.uid = str(uid)
+
+    @staticmethod
+    def create_endpoint_id(function_uri: str, versioned_model: str):
+        function_uri = ModelEndpoint.FunctionURI.from_string(function_uri)
+        versioned_model = ModelEndpoint.VersionedModel.from_string(versioned_model)
+
+        if (
+            not function_uri.project
+            or not function_uri.function
+            or not versioned_model.model
+        ):
             raise ValueError(
-                "ModelEndpoint.spec.function, ModelEndpoint.spec.model "
-                "and ModelEndpoint.metadata.tag must be initalized"
+                "Both function_uri and versioned_model have to be initialized"
             )
 
-        endpoint_unique_string = (
-            f"{self.spec.function}_{self.spec.model}_{self.metadata.tag}"
+        uid = ModelEndpoint.EndpointUID(
+            function_uri.project,
+            function_uri.function,
+            function_uri.tag,
+            versioned_model.model,
+            versioned_model.version,
         )
-        md5_str = md5(endpoint_unique_string.encode("utf-8")).hexdigest()
-        return f"{self.metadata.project}.{md5_str}"
 
+        return uid
+
+    @dataclass
+    class FunctionURI:
+        project: str
+        function: str
+        tag: Optional[str] = None
+        hash_key: Optional[str] = None
+
+        @classmethod
+        def from_string(cls, function_uri):
+            project, uri, tag, hash_key = parse_versioned_object_uri(function_uri)
+            return cls(
+                project=project,
+                function=uri,
+                tag=tag or None,
+                hash_key=hash_key or None,
+            )
 
-class Histogram(BaseModel):
-    buckets: List[Tuple[float, float]]
-    count: List[int]
+    @dataclass
+    class VersionedModel:
+        model: str
+        version: Optional[str]
 
+        @classmethod
+        def from_string(cls, model):
+            try:
+                model, version = model.split(":")
+            except ValueError:
+                model, version = model, None
 
-class Metric(BaseModel):
-    name: str
-    start_timestamp: str
-    end_timestamp: str
-    headers: List[str]
-    values: List[Tuple[str, float]]
-    min: float
-    avg: float
-    max: float
+            return cls(model, version)
 
+    @dataclass
+    class EndpointUID:
+        project: str
+        function: str
+        function_tag: str
+        model: str
+        model_version: str
+        uid: Optional[str] = None
 
-class FeatureValues(BaseModel):
-    min: float
-    avg: float
-    max: float
-    histogram: Histogram
+        def __post_init__(self):
+            versioned_function = f"{self.function}_{self.function_tag or 'N/A'}"
+            versioned_model = f"{self.model}_{self.model_version or 'N/A'}"
+            unique_string = f"{self.project}_{versioned_function}_{versioned_model}"
```

Comment:
1. you're ignoring that function version can be expressed by hash_key, not only tag
2. IIRC "versioned_model" is just a user input to name (and version) the mode **endpoint**, it means that possibly there might be a model (specified by the `model_artifact` attribute) on a function, and two model endpoint records will be created for it just because the user gave different name ?
I mean, on one hand we said what identify a model endpoint is the mix of a specific function + model - which make me feels like we should use here the `model_artifact`, on the other hand - if we do this, I guess we still should ensure uniqueness somehow on the `versioned_model` thing ?

Reply:
1. Does it change anything about this implementation?
2. The only issue i can see with using `model_uri` is that it is optional (true for `model_version` as well), but the name will always be supplied. @yaronha any thoughts on this one?